---
title: "Selecting genes with L-shaped association from expression and methylation arrays"
author: "Berta Miro Cau and Alex Sánchez-Pla"
output: rmarkdown::html_vignette
vignette: >
 %\VignetteIndexEntry{Selecting genes with negative correlation (L-shaped) from expression and methylation arrays}
 \usepackage[utf8]{inputenc} 
 %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
 
# Introduction

This vignette provides some examples on how to use the 'LHeuristic' package functions to select genes with an expression vs methylation scatterplot that have an L-shaped pattern. 

# Input data

The data provided is a subset of the methylation array and the expression array from the colon adenocarcinoma (COAD) from The Cancer Genome Atlas (TCGA). The full dataset is available from the TCGA website (https://tcga-data.nci.nih.gov/docs/publications/tcga/?).

The methylation and expression data are in 2 different files, the columns are the samples (patients) and the rows are the genes. 

```{r}
library(Lheuristic)
library(kableExtra)
library(VennDiagram)

data("TCGAexp")
data("TCGAmet")

```

The package contains some utility functions such as ` checkPairing` that can be used to check that both matrices have the same rows and column number and names.

```{r}
dim(TCGAexp)
ifelse(checkPairing(TCGAexp, TCGAmet), "Match OK", "Check matching")
```

# Data analysis

The Lheuristic package has been created to implement the heuristic algorithm developed by Sanchez et alt. (@cite()) to detect L-shaped scatterplots. However many researchers are used to rely on negative correlation to call a gene "regulated by methylation", so for completeness, both methods are available in the package.

Each methods has various parameters which are optimized to select for the L-shape in scatterplot data, and with a negative correlation.

The functions allow for parameter tunning, in which case, any other selection is possible.

## Correlation method

The _correlation method_ is a naïve approach consisting of calling those genes showing a significant negative correlation between their expression and methylation values.

```{r}
correlation <- correlationSelection (TCGAexp, TCGAmet, pValCutoff=0.25,  rCutoff=-0.5, type="Spearman",adj=TRUE)
correlationL <-correlation[correlation$SigNegCorr,] 
correlationNoL <-correlation[!correlation$SigNegCorr,] 

cat("The number of genes selected with the correlation method is: ", sum(correlationL$SigNegCorr),"\n")
```

We can observe the distribution of the correlation coeficients:

```{r}
plot(density(correlationL[,1]) , main="Significant correlations in TCGA dataset")
```

Depending on the *pvalue* and the *r* cutoff more or less genes will be retained.

The result is a list with the genes slected and not selected in a table format with columns *r coefficient* and the *p-value* for the chosen correlation, the *adjusted p value*, the *distance correlation* and a logical wheter the gene was classified as L-shaped or not according to the parameters set.

```{r}
kableExtra::kable(correlationL, caption = "Genes selected with L-shape with the correlation method")
```

The resulting genes can also be plotted and saved in a PDF file (leaving the file name empty shows the plot in screen). The figure below shows a picture of the first four selected genes.

```{r}
genes2plot <- rownames(TCGAmet) %in% rownames(correlationL)[1:4]
opt <- par(mfrow=c(2,2))
plotGenesMat (mets=TCGAmet[genes2plot,1:4], 
               expres=TCGAexp[genes2plot,1:4], 
               fileName = "correlationLgenes.pdf",
               text4Title = correlationL[rownames(correlationL),""]) 
par(opt)
```


## Heuristic method

The heuristic method intends to select L-shaped scatterplots by superimposing a grid on the graph and defining cells which have to (or don't have to) contain a minimum (or maximum) percentage of points if the scatterplot is to be called L-shaped.

The method also computes a score in such a way that scores in selected regions (L region) score positively and points off the region of interest score negatively. An appropriate setting of scores and weights should yield positive scores for L-shaped scatterplots and negative scores for those that are not. 

```{r}
sampleSize <- dim(TCGAmet)[2]
numGenes <-   dim(TCGAmet)[1]

# 
reqPercentages <- matrix (c(2, 20, 5, 1, 40, 20, 0, 1, 2), nrow=3, byrow=TRUE)
(maxminCounts <- toReqMat(sampleSize, reqPercentages))

(theWeightMifL=matrix (c(2,-2,-sampleSize/5,1,0,-2,1,1,2), nrow=3, byrow=TRUE))
(theWeightMifNonL=matrix (c(0,-2,-sampleSize/5,0,0,-2,0,0,0), nrow=3, byrow=TRUE))
   
   

heur <- scoreGenesMat (mets=TCGAmet,
							              expres=TCGAexp,
                            aReqPercentsMat=reqPercentages,
                            aWeightMifL=theWeightMifL,
                            aWeightMifNonL=theWeightMifNonL )
  cat("Number of scatterplots scored  : ", dim(heur)[1],"\n")
  cat("Number of L-shape scatterplots : ", sum(heur[,1]),"\n")
  
heurL <- heur[heur$logicSc,]
heurNoL <- heur[!heur$logicSc,]
```

We can check the results in the following table, were there is a logical value describing if the gene has or not an L-shape based on our criteria and the *numerSc* score:

```{r}
kableExtra::kable((heurL), caption = "Genes selected with L-shape with the heuristic method")
```

Next, we can visualize the scatterplots of the selected genes or save them on a PDF file.

```{r}
genes2plot2 <- rownames(TCGAmet) %in% rownames(heurL)[1:4]
opt <- par(mfrow=c(2,2))
plotGenesMat (mets=TCGAmet[genes2plot2,], 
               expres=TCGAexp[genes2plot2,], 
               fileName ="selectedGenes.pdf",
               text4Title = heurL[genes2plot2,"numeriSc"]) 
par(opt)
```

# Comparison of selected L-shaped genes

We have lists of genes that we have identified with the 2 different methods. We can create the intersection of all lists and then visualize the results with a Venn Diagram.

```{r}
library(VennDiagram)
myVenn<- venn.diagram(x=list(correlation=rownames(correlationL),
                                heuristic = rownames(heurL)), 
                                filename=NULL, lty = "blank",  
                                fill=c("pink1", "skyblue"),
                             main="Genes in common between the 2 methods")
grid.newpage()
grid.draw(myVenn)
```

We can decide to choose the genes that have been selected by 2 or more methods, for example, to have higher consistancy in the selection. 

```{r}
inCommonL <- intersect(rownames(correlationL), rownames(heurL))
```

We can also plot selected genes.

```{r, fig.height=17, fig.width=12}
par(mfrow=c(2,2))
myGene1 <-inCommonL[1]
xVec<- as.numeric(TCGAmet[myGene1,])
yVec<-as.numeric(TCGAexp[myGene1,])
titleT <- paste (myGene1, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)


myGene2 <-inCommonL[2]
xVec<- as.numeric(TCGAmet[myGene2,])
yVec<-as.numeric(TCGAexp[myGene2,])
titleT <- paste (myGene2, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)


myGene3 <-inCommonL[3]
xVec<- as.numeric(TCGAmet[myGene3,])
yVec<-as.numeric(TCGAexp[myGene3,])
titleT <- paste (myGene3, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)

myGene5 <-inCommonL[5]
xVec<- as.numeric(TCGAmet[myGene3,])
yVec<-as.numeric(TCGAexp[myGene3,])
titleT <- paste (myGene3, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)
```

