---
title: "Selecting genes with L-shaped association from expression and methylation arrays"
author: "Berta Miro Cau and Alex Sánchez-Pla"
date: "`r Sys.Date()`"
output:
  rmarkdown::html_vignette: default
vignette: > 
  %\VignetteIndexEntry{Selecting genes with L-shaped association from expression and methylation arrays}
  %\usepackage[utf8]{inputenc}  
  %\VignetteEngine{knitr::rmarkdown}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```
 
# Introduction

This vignette provides some examples on how to use the `LHeuristic` package functions to select genes with an expression vs methylation scatterplot that have an L-shaped pattern. 

Essentially what a user may want to do uysing this package is:

-   Data input

    -   Select (Load) expression values from a csv file

    -   Select (Load) methylation values from a csv file

    -   Data preprocessing may need to be done automatically so that its results can be used for filtering.

    -   Set parameter values

        -   For data use (all genes or only those that pass certain filters ($\mbox{corr}< 0$))
        -   For L-shape selection method: Here one may want to select the method and its parameters (each method has a different set of parameters)

-   Data processing (For each method and set of parameters ...)

    -   Run the computation and "mark" L-shape genes
    -   Draw the scatterplots of either all genes or only selected genes

-   Data output (For each method and set of parameters ...)

    -   Save (Download) the resulting gene list(s)
    -   Save (Download) the scatterplots

# Input data

The data provided is a subset of the methylation array and the expression array from the colon adenocarcinoma (COAD) from The Cancer Genome Atlas (TCGA). The full dataset is available from the TCGA website (https://tcga-data.nci.nih.gov/docs/publications/tcga/?).

The methylation and expression data are in 2 different files, the columns are the samples (patients) and the rows are the genes. 

```{r}
library(Lheuristic)
# library(kableExtra)
# library(VennDiagram)

data("TCGAexp")
data("TCGAmet")

```

The package contains some utility functions such as `checkPairing` that can be used to check that both matrices have the same rows and column number and names.

```{r}
dim(TCGAexp)
ifelse(checkPairing(TCGAexp, TCGAmet), "Match OK", "Check matching")
```

# Data analysis

The Lheuristic package has been created to implement the heuristic algorithm developed by Sanchez et alt. (@cite()) to detect L-shaped scatterplots. However many researchers are used to rely on negative correlation to call a gene "regulated by methylation", so for completeness, both methods are available in the package.

Each methods has various parameters which are optimized to select for the L-shape in scatterplot data, and with a negative correlation.

The functions allow for parameter tunning, in which case, orther forms of correlation selection are possible.

## Correlation method

The _correlation method_ is a naïve approach consisting of calling those genes showing a significant negative correlation between their expression and methylation values.

```{r}
correlation <- correlationSelection (TCGAexp, TCGAmet, pValCutoff=0.25,  rCutoff=-0.5, type="Spearman",adj=TRUE)
correlationL <-correlation[correlation$SigNegCorr,] 
correlationNoL <-correlation[!correlation$SigNegCorr,] 

cat("The number of genes selected with the correlation method is: ", sum(correlationL$SigNegCorr),"\n")
```

We can observe the distribution of the correlation coeficients:

```{r,fig.width = 6, fig.height=4}
plot(density(correlationL[,1]) , main="Significant correlations in TCGA dataset")
```

Depending on the *pvalue* and the *r* cutoff more or less genes will be retained.

The result is a list with the genes slected and not selected in a table format with columns *r coefficient* and the *p-value* for the chosen correlation, the *adjusted p value*, the *distance correlation* and a boolean value indicating wheter the gene was selected for having a significantly negative correlation.

```{r}
head(correlationL)
```

The resulting genes can also be plotted and saved in a PDF file (leaving the file name empty shows the plot in screen). The figure below shows a picture of the first four selected genes.

```{r, fig.height=6, fig.width=6}
genes2plot <- rownames(TCGAmet) %in% rownames(correlationL)[1:4]
opt <- par(mfrow=c(2,2))
plotGenesMat (mets=TCGAmet[genes2plot,1:4], 
               expres=TCGAexp[genes2plot,1:4], 
               fileName = NULL, #"correlationLgenes.pdf",
               text4Title = correlationL[rownames(correlationL),""]) 
par(opt)
```


To plot the scatterplots depicting the relation between methylation and expression for a list of genes is achieved with the function `plotGenesMat`.

## Heuristic method

The heuristic method intends to select L-shaped scatterplots by superimposing a grid on the graph and defining cells which have to (or don't have to) contain a minimum (or maximum) percentage of points if the scatterplot is to be called L-shaped.

The method also computes a score in such a way that scores in selected regions (L region) score positively and points off the region of interest score negatively. An appropriate setting of scores and weights should yield positive scores for L-shaped scatterplots and negative scores for those that are not. 

This idea can be made more explicit by introducing a "three-band rule" as follows:

1.  Overimpose a $3\times 3$ grid on the scatterplot.

2.  Classify the scatterplot as **"L" or "non-L"** based on a small set of conditions:

    2.1 There must be a *minimum* number of points in the upper-left (cell (1,1)) and lower right (cell (3,3)) corners of the grid.

    2.2 There must be a *maximum* number of points in the upper right (cell (1,3)) because points there mean hypermethylation and hyperexpression which is the opposite of what we are looking for.

    2.3 We will usually *not require to have a minimum of points in cell (3,1)* unless we are really willing to have an L-shape (in our setting we will also be happy tho recover diagonals, which also reflect a negative correlation!).

3.  Score points on each subgrid in such a way that

    3.1 Points in permitted regions (left-outer margin, i.e. cells: (1,1), (2,2), (3,1), (3,2), (3,3)) score positively if the scatterplot has been classified as L or zero if it has been classified as non-L.

    3.2 Points in non-desired regions (outer band. i.e. cells (1,2), (1,3), (2,3)) score negatively in all cases.

    3.3 Some regions may be declared neutral and not-score, such as cell (2,2).

4.  *Use cross-validation to tune scoring parameters* (**if a set of positive and negative L-shaped genes is available**).

The previous scheme can be summarized using the following equation.
$$
S(X) = W_L \circ X \times \mathbf{1}_L(X) + W_{L^C} \circ X \times \mathbf{1}_{L^c}(X),
$$ where

-   ${X}$ is the matrix of *counts*, i.e. the number of counts in each cell of the grid,
-   ${W_L}$ is the matrix of scores per cell and point *if the scatterplot has been classified as* $L$,
-   ${W_{L^c}}$ is the matrix of scores per cell and point *if the scatterplot has been classified as non-*$L$ ($L^c$), and $\circ$ represents the hadamard product of the two matrices $W_{L/L^c}$ (i.e. elementwise multiplication of the two matrices) and $\mathbf{1}_{L/L^c}()$ is the indicator function for $L$ or $L^c$.

The fact that the scatterplot is assigned to $L$ or $L^c$ can also be described as the hadamard product of three matrices: $$
\mathbf{1}_L(X) = \bigwedge_{i,j} X \circ C \circ \left( mMP \times \sum_{i,j}x_{ij}\right),
$$ where
-   ${X}$ is the matrix of *counts*, i.e. the number of counts in each cell of the grid,
-   $C$ is the matrix of conditions to be verified *if the scatterplot has to be classified as* $L$,
-   $mMP$ is the matrix of minimum and Maximum Percentages of points to have in each cell *if the scatterplot has to be classified as* $L$,
-   $\circ$ represents the pointwise logical operation which allows that the product of the three cells becomes a logical operation and
-   $\bigwedge_{i,j}$ represents an logical "AND" operation of all cells, that is if all cells are TRUE the result is assigned to $L$ and if one fails it is assigned to $L^c$.


```{r}
sampleSize <- dim(TCGAmet)[2]
numGenes <-   dim(TCGAmet)[1]

# 
reqPercentages <- matrix (c(2, 20, 5, 5, 40, 20, 3, 3, 2), nrow=3, byrow=TRUE)
sum(reqPercentages)
(maxminCounts <- toReqMat(sampleSize, reqPercentages))

(theWeightMifL=matrix (c(2,-2,-sampleSize/5,1,0,-2,1,1,2), nrow=3, byrow=TRUE))
(theWeightMifNonL=matrix (c(0,-2,-sampleSize/5,0,0,-2,0,0,0), nrow=3, byrow=TRUE))
   
heur <- scoreGenesMat (mets=TCGAmet,
							              expres=TCGAexp,
                            aReqPercentsMat=reqPercentages,
                            aWeightMifL=theWeightMifL,
                            aWeightMifNonL=theWeightMifNonL )
  cat("Number of scatterplots scored  : ", dim(heur)[1],"\n")
  cat("Number of L-shape scatterplots : ", sum(heur[,1]),"\n")
  
heurL <- heur[heur$logicSc,]
heurNoL <- heur[!heur$logicSc,]
```

We can check the results in the following table, were there is a logical value describing if the gene has or not an L-shape based on our criteria and the *numerSc* score:

```{r}
knitr::kable(heurL)
```

Next, we can visualize the scatterplots of the selected genes or save them on a PDF file.

```{r, fig.height=6, fig.width=6}
genes2plot2 <- rownames(TCGAmet) %in% rownames(heurL)[1:4]
opt <- par(mfrow=c(2,2))
plotGenesMat (mets=TCGAmet[genes2plot2,], 
               expres=TCGAexp[genes2plot2,], 
               fileName =NULL, # "selectedGenes.pdf",
               text4Title = heurL[genes2plot2,"numeriSc"]) 
par(opt)
```


# Comparison of selected L-shaped genes

Once we have lists of genes that we have identified with the 2 different methods we can create the intersection of all lists and then visualize the results with a Venn Diagram.

We may decide to choose the genes that have been selected by 2 or more methods, for example, to have higher consistency in the selection. 

```{r}
inCommonL <- intersect(rownames(correlationL),
                       rownames(heurL))
inCorrelationLOnly<- setdiff(rownames(correlationL),
                             inCommonL)
inheurLLOnly<- setdiff(rownames(heurL),
                             inCommonL)
```


We can also plot selected genes.
As an example we will plot the genes 1, 2, 3 and 5 by index.

```{r, fig.height=6, fig.width=6}
par(mfrow=c(2,2))
myGene1 <-inCommonL[1]
xVec<- as.numeric(TCGAmet[myGene1,])
yVec<-as.numeric(TCGAexp[myGene1,])
titleT <- paste (myGene1, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)

myGene2 <-inCommonL[2]
xVec<- as.numeric(TCGAmet[myGene2,])
yVec<-as.numeric(TCGAexp[myGene2,])
titleT <- paste (myGene2, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)


myGene3 <-inCommonL[3]
xVec<- as.numeric(TCGAmet[myGene3,])
yVec<-as.numeric(TCGAexp[myGene3,])
titleT <- paste (myGene3, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)

myGene5 <-inCommonL[5]
xVec<- as.numeric(TCGAmet[myGene3,])
yVec<-as.numeric(TCGAexp[myGene3,])
titleT <- paste (myGene3, "(May be GRM)")
plotGeneSel(xMet=xVec, yExp=yVec, titleText=titleT, x1=1/3, x2=2/3)
```


Essentially the process of selecting L-shape genes consists of three steps:

1.  Select datasets (a pair of row-column matched matrices, one for expression and one for methylation).

2. Set parameters:

2.1 Grid definition

2.2 Binary Scoring

2.3.Numerical Scoring

4.  Score the selected data and return classification (scores and group) and plots for each gene.




